Fix forms with "undefined" xmlns

NoahCarnahan opened the pull request:
Addresses part of http://manage.dimagi.com/default.asp?221525
This should be able to migrate the majority of xforms and forms that have "undefined" as their xmlns.

@emord cc @esoergel @snopoke

NoahCarnahan committed:
Add management command

NoahCarnahan committed:
more work on the management command

dimagimon commented:
There are 28 errors:

* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 120 - E122 continuation line missing indentation or outdented
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 58 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 78 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 51 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 71 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 38 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 30 - E122 continuation line missing indentation or outdented
E901 SyntaxError: invalid syntax
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 52 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 34 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 147 - E122 continuation line missing indentation or outdented
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 97 - E124 closing bracket does not match visual indentation
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 76 - E123 closing bracket does not match indentation of opening bracket's line
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 69 - E122 continuation line missing indentation or outdented
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 36 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 43 - E226 missing whitespace around arithmetic operator
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 32 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 57 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 152 - E123 closing bracket does not match indentation of opening bracket's line
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 77 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 54 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 44 - E226 missing whitespace around arithmetic operator
E501 line too long (119 > 115 characters)
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 17 - E302 expected 2 blank lines, found 1
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 143 - E122 continuation line missing indentation or outdented
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 37 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 83 - E122 continuation line missing indentation or outdented
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 103 - E122 continuation line missing indentation or outdented
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 139 - E122 continuation line missing indentation or outdented
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 33 - E251 unexpected spaces around keyword / parameter equals

dimagimon commented:
There are 28 errors:

* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 120 - E122 continuation line missing indentation or outdented
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 58 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 78 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 51 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 71 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 38 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 30 - E122 continuation line missing indentation or outdented
E901 SyntaxError: invalid syntax
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 52 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 34 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 147 - E122 continuation line missing indentation or outdented
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 97 - E124 closing bracket does not match visual indentation
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 76 - E123 closing bracket does not match indentation of opening bracket's line
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 69 - E122 continuation line missing indentation or outdented
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 36 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 43 - E226 missing whitespace around arithmetic operator
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 32 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 57 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 152 - E123 closing bracket does not match indentation of opening bracket's line
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 77 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 54 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 44 - E226 missing whitespace around arithmetic operator
E501 line too long (119 > 115 characters)
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 17 - E302 expected 2 blank lines, found 1
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 143 - E122 continuation line missing indentation or outdented
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 37 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 83 - E122 continuation line missing indentation or outdented
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 103 - E122 continuation line missing indentation or outdented
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 139 - E122 continuation line missing indentation or outdented
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 33 - E251 unexpected spaces around keyword / parameter equals

NoahCarnahan commented:
This PR is a follow on from https://github.com/dimagi/commcare-hq/pull/11033

NoahCarnahan committed:
fix typo

dimagimon commented:
There are 8 errors:

* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 121 - E303 too many blank lines (3)
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 34 - F841 local variable 'xlmns_map_path' is assigned to but never used
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 45 - E226 missing whitespace around arithmetic operator
E501 line too long (119 > 115 characters)
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 44 - E226 missing whitespace around arithmetic operator
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 17 - E302 expected 2 blank lines, found 1
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 98 - E124 closing bracket does not match visual indentation
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 8 - F401 'term' imported but unused
* corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py, line 10 - F401 'const' imported but unused

NoahCarnahan committed:
lint

NoahCarnahan committed:
remove line that should have been deleted

dimagimon commented on corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py:
```diff
@@ -0,0 +1,157 @@
+            for i, xform_instance in enumerate(submissions):
+                if i % 500 == 0 and i != 0:
+                    print "Progress: {} of {} ({})".format(i, total, round(i / float(total), 2))
+                    print "We can fix {} of {} ({})".format(things_we_can_fix, i, round(things_we_can_fix / float(i), 2))
```
E501 line too long (121 > 115 characters)

NoahCarnahan committed:
lint

NoahCarnahan commented on corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py:
```diff
@@ -0,0 +1,159 @@
+
+
+def get_submissions_without_xmlns():
+    # TODO: This view won't get archived forms etc
```
Is this an issue? I suppose it might be if someone ever unarchives a submissions with "undefined" xmlns

NoahCarnahan commented on corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py:
```diff
@@ -0,0 +1,159 @@
+        key="undefined",
+        include_docs=False,
+        reduce=False,
+    ).all()
```
@emord I tried doing an ES query aggregated by app_id but es aggregations don't return the docs themselves, and the results are approximate

esoergel replied:
I do think it's best to pull the docs from couch since they'll be resaved in couch (there are sometimes subtle differences between ES and couch, like additional fields).  You could get the doc IDs from ES, then pass to `iter_docs` to get the actual documents, but since this view seems to do what you want anyways, it seems like an appropriate course of action.

NoahCarnahan commented on corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py:
```diff
@@ -0,0 +1,159 @@
+
+    form.source = new_xml
+    if not dry_run:
+        # TODO: Alter the builds too?
```
I think we would want to alter the builds too. If we don't a user could revert an app and end up with an "undefined" xmlns again.

NoahCarnahan commented on corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py:
```diff
@@ -0,0 +1,159 @@
+    return 'http://openrosa.org/formdesigner/{}'.format(uuid.uuid4())
+
+
+def replace_xml(xform, new_xml):
```
This function mirrors `XFormInstance.get_xml()`. The attachment can be stored in a few different ways, but I expect all submissions with this issue to be recent, and therefore they should all use the `get_attachment()`/`put_attachment()` storage method

NoahCarnahan committed:
Update xmlns on app builds too

esoergel commented on corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py:
```diff
@@ -0,0 +1,175 @@
+                 .filter(NOT(xmlns('undefined'))))
+        if query.count():
+            return True
+    return False
```
You could probably memoize this function, as I imagine the affected forms will end up calling this with the same arguments a bunch of times.

NoahCarnahan replied:
ah, good call. Will updated.

esoergel commented on corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py:
```diff
@@ -0,0 +1,175 @@
+                    forms_without_xmlns = get_forms_without_xmlns(app)
+                    if len(forms_without_xmlns) == 1:
+                        form = forms_without_xmlns[0]
+                        if not xforms_with_real_xmlns_possibly_exist(app._id, form):
```
Should there be an `else` clause here?  Or is adding in the real xmlnss a separate step?

NoahCarnahan replied:
Should have been more explicit about that, sorry. Currently not planning on attempting to fix forms where `xforms_with_real_xmlns_possibly_exist()` is true, because there could be false positives, and I haven't thought through how exactly to determine if these form matches are real or not. I'll either try to match by questions, or even look at them manually if the number is low enough.

esoergel commented on corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py:
```diff
@@ -0,0 +1,175 @@
+    xform_instance.xmlns = xmlns
+    xform_instance.form['@xmlns'] = xmlns
+    if not dry_run:
+        xform_instance.save()
```
Do you have an order-of-magnitude estimate of how many forms are affected?  If it's massive, it could be worth looking into saving in chunks.

NoahCarnahan replied:
I believe it's about 64k, so not small, but probably worth bulk saving. I'll update it.

esoergel replied:
One approach to that would be to use `with IterDB(XFormInstance.get_db()) as db:` in the `handle` method, then just pass around the `db` instance.

NoahCarnahan replied:
(updated)

NoahCarnahan committed:
memoize xforms_with_real_xmlns_possibly_exist()

NoahCarnahan committed:
Fix error xforms (like archived forms) too

NoahCarnahan committed:
Improve xforms_with_real_xmlns_possibly_exist()

Check error xforms (like archived forms) in addition to normal
xforms when looking for possible preexisitng xmlns.

NoahCarnahan committed:
[refactor] Pull print progress into separate function

NoahCarnahan committed:
Use IterDB for xform saves

NoahCarnahan committed:
Improve exception handling

NoahCarnahan committed:
Log new xmlns for each build

NoahCarnahan committed:
line length

NoahCarnahan committed:
Use IterDB from apps

NoahCarnahan committed:
Only save builds that have the undefined xmlns

NoahCarnahan committed:
Remove unnecessary try/except

NoahCarnahan committed:
log failed saves

dimagimon commented on corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py:
```diff
@@ -0,0 +1,234 @@
+                f.write("Failed to save {}\n".format(error_id))
+
+
+    def _print_progress(self, i, total_submissions, num_fixed):
```
E303 too many blank lines (2)

NoahCarnahan committed:
lint

NoahCarnahan commented:
I'm thinking I want to run this management command before coding a migration for the harder to solve cases, as this command as is should be able to migrate the majority of xforms.
Anyone want to take a final pass?
(@czue might also be interested)

czue commented on corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py:
```diff
@@ -0,0 +1,233 @@
+                            )
+                            f.write("Set new xmlns on submission {}\n".format(xform_instance._id))
+                        else:
+                            app = Application.get(xform_instance.app_id)
```
it may not matter in this case since this was a bug in vellum, but using `get_app` is preferred if you don't know the app source in case it comes from a remote app

czue commented on corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py:
```diff
@@ -0,0 +1,233 @@
+                                        xform_db,
+                                        dry_run,
+                                    )
+                                    f.write("Set new xmlns on submission {}\n".format(xform_instance._id))
```
if something goes wrong you may find yourself wishing you had also logged `new_xmlns` here

NoahCarnahan replied:
a4015a2

czue commented on corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py:
```diff
@@ -0,0 +1,233 @@
+                            forms_without_xmlns = get_forms_without_xmlns(app)
+                            if len(forms_without_xmlns) == 1:
+                                form = forms_without_xmlns[0]
+                                if not xforms_with_real_xmlns_possibly_exist(app._id, form):
```
This is really an edge case, but do we not also have to worry about the case where the app had multiple `undefined` xmlnses and then one was fixed? e.g. by also comparing the form name in the submission with the app and making sure it looks right.

emord replied:
Hmm. I don't think I thought about that one. There is a validation on saving the xform that forces the xmlns to be unique within the app, but if one was fixed in the time that the bug was live that could cause another form to have an `undefined` xmlns. Sounds like comparing form name would be a good idea.

NoahCarnahan replied:
Sure I can compare the name too. Just to make sure I'm understanding the issue correctly, is this what you had in mind:

- App has Form A and Form B
- Form A saved with "undefined" xmlns
- Submissions are made against Form A
- Form A had xmlns set to a uuid (by someone manually?)
- Form B saved with "undefined" xmlns
- Submissions are made against Form B
- This management command is run, and the submissions against Form A are mistakenly given the new xmlns for Form B

emord replied:
that sounds right to me. I don't think we want to fix the Form A in this case, but just want to flag it as an error

czue replied:
yeah that sounds right to me too, though I think the simpler version is:

- App gets form A and B both with "undefined" xmlns
- Submissions are made to both.
- One of the two is manually fixed (or more likely, deleted from the app)

Might be easier to use the build IDs from the submissions to do the checks against the app (because then you know there was only one undefined form from the build the form was submitted from) and then if those forms are still in the app (by app builder guid) then you can link them. 

(sorry thinking quickly so not sure that all make sense)

czue replied:
we could even consider fixing in the builds in case people revert, but that could be overkill... 

do we have a sense how widespread this problem is?

NoahCarnahan replied:
fixed here: 4f937f1

emord replied:
360 apps over 270 domains. There's a tsv file in the linked ticket. It's not really noticeable at all because no apps had two xforms with undefined xmlns at the same time thanks to a check when you save a form in HQ, so all reports/exports should still work

NoahCarnahan replied:
Oh right the build_id definitely sounds better for disambiguating the forms.

Already modifying the forms in all builds here: https://github.com/dimagi/commcare-hq/blob/eb09592cfdf89bbccc16bccda022335cb427cb64/corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py#L188

czue commented on corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py:
```diff
@@ -0,0 +1,233 @@
+    return False
+
+
+def set_xmlns_on_submission(xform_instance, xmlns, xform_db, dry_run):
```
I think it might be good if this function also saved a property to the form for auditing purposes just in case we ever need to go back and figure out something in the data as a result of this migration doing something unexpected. I think that could just be a property called `form_migrated_from_undefined_xmlns` and a boolean or a date or something.

NoahCarnahan replied:
good thinking. 2d619a2

czue commented:
wow this looks like a nasty problem. overall looks great, had some minor comments. could also be nice to have some tests, though understand it's a one-off.

emord commented:
@NoahCarnahan Thanks for taking this over! Looks good to me pending @czue comments

NoahCarnahan committed:
Check xform name against form name before updating xmlns

NoahCarnahan committed:
Add form_migrated_from_undefined_xmlns field to models

NoahCarnahan committed:
Log new xmlns for every xform

NoahCarnahan committed:
Use get_app

dimagimon commented on corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py:
```diff
@@ -0,0 +1,246 @@
+                                    if not xforms_with_real_xmlns_possibly_exist(app._id, form):
+                                        num_fixed += 1
+                                        new_xmlns = generate_random_xmlns()
+                                        new_xmlnss[xform_instance.app_id] = NewXmlnsInfo(form.name.values(), new_xmlns)
```
E501 line too long (119 > 115 characters)

NoahCarnahan committed:
add tests

dimagimon commented on corehq/apps/cleanup/tests/test_fix_forms_with_missing_xmlns.py:
```diff
@@ -0,0 +1,188 @@
+
+DOMAIN = "test"
+
+class TestFixFormsWithMissingXmlns(TestCase, TestXmlMixin):
```
E302 expected 2 blank lines, found 1

NoahCarnahan committed:
lint

czue commented:
@NoahCarnahan sorry it took so long to get here but given @emord's point that no build can have multiple forms with `"undefined"` at the same time I think that we can do a clear resolution that reliably catches every single instance of this with the following process:

Fixing the forms:

- From the form, get the build ID and the app ID
- From the build, get the [form `unique_id`](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/app_manager/models.py#L753) in app manager associated with the form
- (assert the form names are the same - should be impossible to be otherwise)
- Get the XMLNS for that form unique_id - either by looking it up from a dict that maps `unique_id` to `xmlns` or by generating a new one
- Update and save the form
- Add the app ID to a dict mapping apps to process to sets of form unique_ids to fix, if not already there
- Add the form `unique_id` to the set of app forms to fix, if not already there.

Fixing the apps:

- Loop through the list of apps to fix
- Loop through the builds of that app
- For each build, update all occurrences of form unique_ids present in that build accordingly and save.

Does that sound right? And do you think it'd be worth updating this so that's what it does? (I think 90% of the work is done/the same and just a few changes needed to the outer processing code)

emord commented:
That looks right to me. For some reason I didn't see `build_id` in the xform model and I thought you wouldn't be able to get it.

After this gets run we should email the affected projects and ask them to redeploy their apps. 

Until all of their apps are redeployed we'll need to keep an eye on any new submissions in those domains and switch them over. (I can do that)

NoahCarnahan commented:
Hey, that does sound right, and I think it is worth updating this management command to do that. One thing:

> Get the XMLNS for that form unique_id - either by looking it up from a dict that maps unique_id to    xmlns or by generating a new one

I'm imagining that this dict is populated as we generate new ids. I think if the unique_id is not in the dict, we also need to check to see if this unique_id ever had a real xmlns associated with it (such as if this form was fixed manually). So I'm thinking this looks something like this:
```python
if form.unique_id not in unique_id_to_xmlns_map:
   existing_xmlns = set()
   for build in [app] + get_saved_apps(app):
      form_in_build = app_build.get_form(form.unique_id)
      if form_in_build.xmlns != "undefined":
         existing_xmlns.add(form_in_build.xmlns)
   assert len(existing_xmlns) == 0 or len(existing_xmlns) == 1:
   unique_id_to_xmlns_map[form.unique_id] = existing_xlmns.pop() if len(existing_xmlns) else generate_new_xmlns()
```
Does that seem right?

czue commented:
@NoahCarnahan good call, that definitely seems right.

NoahCarnahan committed:
rewrite

dimagimon commented on corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py:
```diff
@@ -0,0 +1,268 @@
+    assert len(existing_xmlns) == 0
+    return generate_random_xmlns()
+
+def _name_matches(xform_name, form_names):
```
E302 expected 2 blank lines, found 1

dimagimon commented on corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py:
```diff
@@ -0,0 +1,268 @@
+                i, total_submissions, round(i / float(total_submissions), 2)
+            )
+
+def get_submissions_without_xmlns():
```
E302 expected 2 blank lines, found 1

dimagimon commented on corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py:
```diff
@@ -0,0 +1,268 @@
+        )
+    return form.unique_id
+
+def get_xmlns(form_unique_id, app_id, domain):
```
E302 expected 2 blank lines, found 1

dimagimon commented on corehq/apps/cleanup/tests/test_fix_forms_with_missing_xmlns.py:
```diff
@@ -0,0 +1,161 @@
+        return good_form, bad_form, good_xform, bad_xforms
+
+
+    def build_app_with_recently_fixed_form(self):
```
E303 too many blank lines (2)

NoahCarnahan committed:
lint

NoahCarnahan committed:
[ci skip] remove comment

NoahCarnahan commented:
@czue @emord 
So... Looks like submissions made in cloudcare preview mode don't have a `build_id`.

We could skip them on the first pass, and if the app corresponding to the `xform.app_id` only has one `form.unique_id` for which the xmlns needs to be fixed, then we could assume that these submissions belong to that form. Various other things we could do to try to guess at the right form. Or, we could just skip fixing them altogether...
Looks like there are 183 xforms in this situation (in 66 domains)

czue commented:
given that those are definitely fake data submitted by web users I'd opt for whatever the easiest/best thing you think is right now that isn't too much effort

NoahCarnahan commented:
Ok, I'm in favor of not attempting to fix those cloudcare forms then. I think this is ready to merge (and run!).

czue commented:
:clap: :clap:
